File: Medium/59.螺旋矩阵-ii.cpp
/*
 * @lc app=leetcode.cn id=59 lang=cpp
 *
 * [59] 螺旋矩阵 II
 */

// @lc code=start
class Solution
{
    //和第五十四题螺旋矩阵一样的思路
public:
    vector<vector<int>> generateMatrix(int n)
    {
        //用于返回的vector，它是一个n*n的矩阵
        vector<vector<int>> res(n, vector<int>(n, 0));
        //特殊情况
        if (n == 0)
            return res;
        //和第五十四题螺旋矩阵一样的思路
        int num = 1;
        int left = 0, right = n - 1;
        int up = 0, down = n - 1;
        //边界大小决定是否循环
        while (left <= right && up <= down)
        {
            //从左往右填
            for (int i = left; i <= right; i++)
            {
                res[up][i] = num++;
            }
            //一行填完更新上边界
            up++;
            //边界在判断一下
            if (up > down)
                break;
            //从上往下填
            for (int i = up; i <= down; i++)
            {
                res[i][right] = num++;
            }
            //一列填完更新右边界
            right--;
            //边界在判断一下
            if (right < left)
                break;
            //从右往左填
            for (int i = right; i >= left; i--)
            {
                res[down][i] = num++;
            }
            //一行填完更新下边界
            down--;
            //从下往上填
            for (int i = down; i >= up; i--)
            {
                res[i][left] = num++;
            }
            //一列填完更新左边界
            left++;
        }
        return res;
    }
};
// @lc code=end

File: Medium/59.螺旋矩阵-ii.go
/*
 * @lc app=leetcode.cn id=59 lang=golang
 *
 * [59] 螺旋矩阵 II
 */

// @lc code=start
// 和56题类似
func generateMatrix(n int) [][]int {
	res := [][]int{}
	for i := 0; i < n; i++ {
		res = append(res, make([]int, n))
	}
	size := n * n
	// 左右边界
	left, right := 0, n-1
	// 上下边界
	up, down := 0, n-1
	num := 1
	for num <= size {
		// 从左到右填充，行不变，列变化
		for i := left; i <= right; i++ {
			res[up][i] = num
			num++
		}
		// 上边界收缩
		up++
		// 从上到下填充，列不变，行变化
		for i := up; i <= down; i++ {
			res[i][right] = num
			num++
		}
		// 右边界收缩
		right--
		// 从右往左填充，行不变，列变化
		for i := right; i >= left; i-- {
			res[down][i] = num
			num++
		}
		// 下边界收缩
		down--
		// 从下到上填充，列不变，行变化
		for i := down; i >= up; i-- {
			res[i][left] = num
			num++
		}
		// 左边界收缩
		left++
	}
	return res
}

// @lc code=end

